### PR TITLE
Add support for general endpoint configuration for s3, to support more s3-api vendors

### PR DESCRIPTION
**Pull request by tomlarkworthy**
fixes https://github.com/markusahlstrand/cloudworker-proxy/issues/123 and https://github.com/markusahlstrand/cloudworker-proxy/issues/126

Added more endpoint configuration for s3 provider.
Switch to `bun test` to run test, enabling removal of mocha as a dependancy, and bootstrap.js to inject worker shims (bun runtime is more like CF workers by default)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "lint": "eslint src",
     "test": "npm run unit && npm run lint",
-    "unit": "bun run mocha test/**/*.test.js --require test/bootstrap.js",
```
Comment (tomlarkworthy) [inline package.json:9]: wasn't running the TS file, lol, actually I am not sure the bun transpiler is properly working yet. Will follow up with that.

File: package.json [hunk 2]
```diff
@@ -35,6 +35,8 @@
     "shortid": "2.2.16"
   },
   "devDependencies": {
+    "@types/mocha": "^10.0.1",
+    "@types/node": "^20.5.9",
```
Comment (tomlarkworthy) [inline package.json:39]: required for the typeschecker on s3 test to work properly

**Comment by tomlarkworthy** [comment]
I think the "build (14.x)" is stuck because thats the old workflow. Maybe my previous merge did not delete it or its configured as a required step that no longer exists?

File: package.json (hunk 1 above)
Comment (markusahlstrand) [reply package.json:9]: I tested adding this as a build script which seems to be working: 
```build": "bun build ./src/index.ts --outdir ./build```

It doesn't output the d.ts files though.. I'll see if I can read up a bit on bun..

Comment (tomlarkworthy) [reply package.json:9]: I think you might configure the .d.ts export behaviour in a tsconfig.json, it definately will read one if there is one around.

**Comment by tomlarkworthy** [comment]
turns out it was super easy to get rid of mocha and just use `bun test`

File: package.json (hunk 1 above)
Comment (tomlarkworthy) [reply package.json:9]: I switched to bun test, it seems to run it properly without config or bootstrap.js

**Comment by tomlarkworthy** [comment]
Note the CI runs all the test https://github.com/markusahlstrand/cloudworker-proxy/actions/runs/6076062838/job/16483444708#step:5:119

File: package.json [hunk 3]
```diff
@@ -44,7 +46,6 @@
     "eslint-plugin-import": "2.22.1",
     "eslint-plugin-prettier": "3.1.4",
     "fetch-mock": "9.11.0",
-    "mocha": "10.2.0",
```
Comment (tomlarkworthy) [inline package.json:47]: another dependancy dead!

File: package.json [hunk 4]
```diff
@@ -35,6 +35,8 @@
     "shortid": "2.2.16"
   },
   "devDependencies": {
+    "@types/mocha": "^10.0.1",
+    "@types/node": "^20.5.9",
     "bun": "^0.8.1",
     "chai": "4.2.0",
```
Comment (tomlarkworthy) [inline package.json:41]: we probably don't need this one either as bun has jest expect syntax built in.

Comment (markusahlstrand) [reply package.json:41]: Feel free to remove :)

File: src/handlers/s3.js [hunk 5]
```diff
@@ -1,15 +1,39 @@
...
+  endpoint,
+  forcePathStyle,
+}) {
   const aws = new AwsClient({
     accessKeyId,
     region,
     secretAccessKey,
   });
 
+  let resolvedEndpoint; // See https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html
```
Comment (markusahlstrand) [inline src/handlers/s3.js:18]: Think this looks good. If we were to break this out to a function we could make an early exit instead? So something like:
```
function getEndpoint(...) {
  if (endpoint && forcePathStyle) {
    const url = new URL(endpoint);
    return`${url.protocol}//${url.host}/${bucket}`;
  }
  
  ... other if's
```

**Review by markusahlstrand — APPROVED** [review]
Looks good!

File: src/handlers/s3.js (hunk 5 above)
Comment (tomlarkworthy) [reply src/handlers/s3.js]: done

**Comment by markusahlstrand** [comment]
It looks good!  I'll go ahead and merge this and have a look at the other PR's

**Pull request merged**